Implement scoped pointers for gdal handled objects

nyalldawson opened the pull request:
Implements a c++11 unique_ptr with custom deleter for storing/automatic clean up of some gdal object types.

This fixes at least 2 leaks I noticed while implementing this, and I suspect it has the potential to save many more...

nyalldawson committed:
Implement scoped pointers for gdal handled objects

m-kuhn commented:
Sounds good. What's left to do? I guess sqlite is another candidate.

nyalldawson commented:
There's many more gdal objects which could benefit from these. Sqlite would be a great one too - I think we leak a lot with those.

rouault commented on src/providers/ogr/qgsogrprovider.cpp:
```diff
@@ -1327,28 +1323,28 @@ bool QgsOgrProvider::addFeaturePrivate( QgsFeature &f, Flags flags )
       OGR_F_SetFieldNull( feature, ogrAttId );
 #else
-      OGR_F_UnsetField( feature, ogrAttId );
+      OGR_F_UnsetField( feature.get(), ogrAttId );
```
missing .get() in #ifdef OGRNullMarker case

rouault commented on src/providers/ogr/qgsogrprovider.cpp:
```diff
@@ -1728,7 +1721,7 @@ bool QgsOgrProvider::changeAttributeValues( const QgsChangedAttributesMap &attr_
         OGR_F_SetFieldNull( of, f );
 #else
-        OGR_F_UnsetField( of, f );
+        OGR_F_UnsetField( of.get(), f );
```
missing .get() in #ifdef OGRNullMarker case

rouault reviewed (changes requested):
LGTM except the 2 cases I outlined

nyalldawson committed:
Fix build

nyalldawson commented:
>  What's left to do? I guess sqlite is another candidate.

Another good candidate: the gdal CPL objects. I suspect we leak a lot of those.

nyalldawson committed:
More gdal scoped pointer types

nyalldawson committed:
Fix build, tests

nyalldawson committed:
Fix crash when creating new rasters

nyalldawson commented:
@rouault how's d024910 for you?

rouault commented:
> @rouault how's d024910 for you?

I've let a few comments.

nyalldawson commented:
> I've let a few comments.

Thanks - much appreciated!

nyalldawson committed:
Explicitly close gdal dataset handles as early as possible

nyalldawson committed:
Safer GDAL delete and close of datasets, which doesn't leak
and works on Windows

See
https://github.com/qgis/QGIS/commit/d024910490a39e65e671f2055c5b6543e06c7042#commitcomment-25194282

nyalldawson committed:
Fix doxygen warning
